### PR TITLE
[RHELC-1275, RHELC-1516] Exit with 1 when rollback fails

### PR DESCRIPTION
**Pull request by hosekadam**
There is a need to exit with exit code 1 when rollback fails to provide the ability to better recognize those types of failures for Insights. This situation can mainly happen when user uses analyze command or during the conversion answers "No" and user doesn't want to continue over point of no return. In those situations, rollback is processed and when it fails, the system is in an unknown state.

**Expected behavior:**
When rollback fails, c2r exits with 1, warning about unknown state and system should be restored from backup is printed. Any summaries of the report aren't printed or saved - because there is any big value of the report when the rollback failed.

**How it was solved:**
The new property `backup_failed` of BackupController was added. This property is based on instance variable `_rollback_failed` which is set when restore method of some restorable fails. 

I chose this solution because it keeps the code simple. Other solution with returning some value required a lot of code changed and added complexity to the codebase.

I also added a new function in `main.py` for printing info after rollback, which we do in two places. The goal was to reduce duplicity in this part of code.

To have captured exceptions during the rollback, I've also removed handling of exceptions in each restorable and let it up to `BackupController`. By this change I found a bug in the code, when the restorable file was removed from the backup folder when doing the restore for conversion purposes.  The file should stay in the backup folder, if we would remove it, then we won't have anything when processing the rollback.

Jira Issues: [RHELC-1275](https://issues.redhat.com/browse/RHELC-1275), [RHELC-1516](https://issues.redhat.com/browse/RHELC-1516)

Checklist

- [x] PR has been tested manually in a VM (either author or reviewer)
- [ ] Jira issue has been made public if possible
- [x] `[RHELC-]` is part of the PR title 
- [ ] GitHub label has been added to help with Release notes 
- [ ] PR title explains the change from the user's point of view
- [ ] Code and tests are documented properly
- [ ] The commits are squashed to as few commits as possible (without losing data) 
- [ ] When merged: Jira issue has been updated to `Release Pending` if relevant

### CODE REVIEW

File: convert2rhel/backup/__init__.py
Comment (r0x0d) [inline convert2rhel/backup/__init__.py:90]: In this case you don't need the variable name here, you should use the type, I believe: 

```suggestion
        :returns list[RestorableChange]: List of RestorableChange objects that were processed by the Controller.
```

File: convert2rhel/backup/__init__.py
Comment (r0x0d) [inline convert2rhel/backup/__init__.py:51]: @hosekadam, do you mind updating the class docstring to include those two properties? 

Just know I realized that we don't have a description even for `_restorables`, so it would be good to update it and include both of them.

File: convert2rhel/backup/__init__.py
Comment (r0x0d) [inline convert2rhel/backup/__init__.py:123]: A docstring here as well, please!

File: convert2rhel/main.py
Comment (r0x0d) [inline convert2rhel/main.py:163]: This will return `False` instead of 1, right?

```
/tmp> python t.py                                                                                                                             03/22/24 09:27:19 AM
/tmp> $env.LAST_EXIT_CODE                                                                                                                     03/22/24 09:27:20 AM
0
/tmp> cat t.py                                                                                                                                03/22/24 09:27:23 AM
def test():
    return False

test()
```

Comment (r0x0d) [reply convert2rhel/main.py:163]: And if it is 0, why not keeping it simply as 0?

File: convert2rhel/main.py
Comment (r0x0d) [inline convert2rhel/main.py:348]:
```suggestion
    elif not pre_conversion_results:
```

File: convert2rhel/main.py
Comment (r0x0d) [reply convert2rhel/main.py:163]: Ah, this test is false... It will always return the exit code for the interpreter instead of the return code of the script. To make it return the exit code for the script, we would need to use sys.exit() passing the return value for the function to it

File: convert2rhel/backup/__init__.py
Comment (hosekadam) [reply convert2rhel/backup/__init__.py]: I will do that. Just I'll wait a bit till the PR will be close to merging

File: convert2rhel/main.py
Comment (SpyTec) [inline convert2rhel/main.py:385]: Blocker. This is out of scope of the PR, please open a new one for this refactor and technical debt fixing :pray:

File: convert2rhel/main.py
Comment (SpyTec) [inline convert2rhel/main.py:175]: We could simply `return backup.backup_control.rollback_failed`

File: convert2rhel/backup/__init__.py
Comment (SpyTec) [inline convert2rhel/backup/__init__.py:131]: Since this is a list I'd prefer changing it to [`any()`](https://docs.python.org/3/library/functions.html#any)
```suggestion
        return any(self._rollback_failures)
```

File: convert2rhel/backup/files.py
Comment (SpyTec) [inline convert2rhel/backup/files.py:186]: `finally` is called even when an exception is raised. How will this affect this overall as this now sets `self.enabled = False`?

File: convert2rhel/backup/files.py
Comment (SpyTec) [inline convert2rhel/backup/files.py:189]: Making this be handled higher up the chain is fine. LGTM!

File: convert2rhel/main.py
Comment (r0x0d) [reply convert2rhel/main.py:175]: I asked @hosekadam to turn return explicitly here to make it easier to look at the return value and see what code we expect. 

But, regardless, with https://issues.redhat.com/browse/RHELC-1276, I'm planning to introduce an enum to match the return codes we (will) have to figure out what does 0/1/2/... means when returning.

File: convert2rhel/backup/__init__.py
Comment (r0x0d) [reply convert2rhel/backup/__init__.py:131]: I think this will have the same effect here, no? Both `any()` and `bool()` would return true in case of having at least one element in the list. Especially because the list is a list[str] and not a list[bool]. 

In any case, not against changing to use `any` here, just think the idea will be the same with `bool` and `any`. 

On the other hand, I would expect seeing an `any` check here in case we would have something like `[False, False, True]`, in this case, `any` would be applicable perfectly here, as we care about the only one `True` and not the others `False` inside the list.

Comment (SpyTec) [reply convert2rhel/backup/__init__.py:131]: For sure seems to be the same. I don't have to think "how does arrays get evaluated as a bool?" when I see an `any()` so I prefer that. Just personal preference for readability

File: convert2rhel/main.py
Comment (SpyTec) [reply convert2rhel/main.py:385]: Checked again and missed the `rollback_failed` check in the initial review, thus thinking this was just a refactor that was unrelated

File: convert2rhel/main.py
Comment (SpyTec) [inline convert2rhel/main.py:364]: I don't like the name too much as we not only do logging in it but also create a report summary. Not sure of the best name tbh

File: convert2rhel/backup/files.py
Comment (hosekadam) [reply convert2rhel/backup/files.py:186]: Yes, you are right, there is no need to have the `finally`. Also, thankfully to this, I found other places where some exception in the `restore` method would be handled and haven't been sent to the `BackupController`

**Comment by hosekadam** [comment]
What was done:
- renamed the `print_info_after_rollback` -> `provide_status_after_rollback`
- applied the suggestions (use the `any()` and improve comment)
- removed exception handling for `RestorablePEMCert` and `RestorableFile` (will be handled in `BackupController`)
- fixed unit tests (mainly removing the tests for exception handling)

File: convert2rhel/main.py
Comment (hosekadam) [reply convert2rhel/main.py]: Renamed to `provide_status_after_rollback`

File: convert2rhel/backup/certs.py
Comment (SpyTec) [inline convert2rhel/backup/certs.py:191]: Can remove the else here. Codecov also claims this isn't tested

File: convert2rhel/backup/files.py
Comment (SpyTec) [inline convert2rhel/backup/files.py:123]: For good measure lets add this to the docstring. That this function throws OSError, IOError. Ask in team chat for syntax if you need it. @r0x0d probably knows if you can't find it

File: convert2rhel/backup/files.py
Comment (SpyTec) [inline convert2rhel/backup/files.py:178]: Same here, mention OSError, IOError in docstring

File: convert2rhel/main.py
Comment (SpyTec) [inline convert2rhel/main.py:170]:
```suggestion
        provide_status_after_rollback(pre_conversion_results, include_all_reports=True)
```

File: convert2rhel/main.py
Comment (SpyTec) [inline convert2rhel/main.py:224]:
```suggestion
        provide_status_after_rollback(pre_conversion_results, include_all_reports=(toolopts.tool_opts.activity == "analysis"))
```
or since it's long
```suggestion
        include_all_reports = toolopts.tool_opts.activity == "analysis"
        provide_status_after_rollback(pre_conversion_results, include_all_reports)
```

File: convert2rhel/main.py
Comment (SpyTec) [inline convert2rhel/main.py:364]:
```suggestion
    """Print after-rollback messages and determine if there is a report to print, or
    if report shouldn't be printed after failure in rollback."""
```

**Review by SpyTec — COMMENTED** [review]
Seems good, minor things. Will you add integration tests?

File: convert2rhel/backup/files.py
Comment (r0x0d) [reply convert2rhel/backup/files.py:123]: It should be: 

```python
"""
<title>

<description>
:raises OSError: why it raises oserror
:raises IOError: Usually, same as the above ^ 
"""
```

Comment (r0x0d) [reply convert2rhel/backup/files.py:123]: Thinking more about this now, maybe we should even consider adding a `warning` in the docstring to mention that whatever exception raised here will bubble up and be handled by the caller of this plugin. 

```python
"""
<title>

<description>

     .. warning::
         Add the warning message here

:raises OSError: why it raises oserror
:raises IOError: Usually, same as the above ^ 
"""
```

Comment (SpyTec) [reply convert2rhel/backup/files.py:123]: Agreed, sounds good!

File: convert2rhel/backup/certs.py
Comment (hosekadam) [reply convert2rhel/backup/certs.py:191]: I think I can't remove the else path there. Because above, there is a test if [`err.errno == errno.ENOENT`](https://github.com/oamg/convert2rhel/blob/main/convert2rhel/backup/certs.py#L186) and if yes, info is just printed about there is nothing to be removed and in that case is no need to raise the exception.

So I'll have to add the unit tests for this line.

File: convert2rhel/backup/files.py
Comment (hosekadam) [reply convert2rhel/backup/files.py:123]: Thanks! I'll add those comment improvements

File: convert2rhel/backup/certs.py
Comment (r0x0d) [reply convert2rhel/backup/certs.py:191]: You could just: 

```python
except OSError as err:
            if err.errno == errno.ENOENT:
                # Resolves RHSM error when removing certs, as the system might not have installed any certs yet
                loggerinst.info("No certificates found to be removed.")
                return
                
          loggerinst.error("OSError({0}): {1}".format(err.errno, err.strerror))
          # Will be handled in BackupController
          raise err
```

Comment (r0x0d) [reply convert2rhel/backup/certs.py:191]: Except with the crappy formatting, but you could return in that if statement.

Comment (hosekadam) [reply convert2rhel/backup/certs.py:191]: Oh! Right, thanks :)

**Comment by hosekadam** [comment]
What was done:
- added name of arg to calls of provide_status_after_rollback and fixed comment in it
- changed the re-raising of the exception in RestorableCert
- added the exceptions into the doc strings
- added unit test for the exception raised in cert.restore
- added exception handling of system_release file for restoring. No unit tests for this, waiting for approval if it's fine.

File: convert2rhel/subscription.py
Comment (SpyTec) [inline convert2rhel/subscription.py:204]: Don't wait for approval to do unit tests, mention what you want to do in chat or elsewhere if you want to double check if it's good. Overall this try catch is fine

File: convert2rhel/backup/__init__.py
Comment (SpyTec) [inline convert2rhel/backup/__init__.py:128]:
```suggestion
        :returns: bool True if the rollback failed, False if rollback succeeded.
```

File: convert2rhel/backup/files.py
Comment (SpyTec) [inline convert2rhel/backup/files.py:114]:
```suggestion
        :arg rollback: bool value to decide if there is need print the rollback messages.
            This argument can also be used during conversion for restoring some file needed
            for conversion and thus won't need rollback messages.
```

File: convert2rhel/backup/files.py
Comment (SpyTec) [inline convert2rhel/backup/files.py:187]:
```suggestion
        """ "
        Remove the file if it was created during conversion.

            .. warning::
                Exceptions are not handled and left for handling by the calling code.

        :raises OSError: When the removal of the file fails.
        :raises IOError: When the removal of the file fails.
        """
```

File: convert2rhel/backup/files.py
Comment (SpyTec) [inline convert2rhel/backup/files.py:120]:
```suggestion
        :raises OSError: When the backed up file is missing.
        :raises IOError: When the backed up file is missing.
```

File: convert2rhel/subscription.py
Comment (SpyTec) [inline convert2rhel/subscription.py:201]:
```suggestion
                description="Failed to restore the /etc/os-release file needed for subscribing the system.",
                diagnosis="The restore failed with error %s." % (str(e)),
```

File: convert2rhel/subscription.py
Comment (SpyTec) [inline convert2rhel/subscription.py:195]:
```suggestion
            loggerinst.critical_no_exit(
                "Failed to restore the /etc/os-release file needed for subscribing the system with message: %s" % str(e)
```

File: convert2rhel/unit_tests/main_test.py
Comment (SpyTec) [inline convert2rhel/unit_tests/main_test.py:141]: This call only tests the logs atm. Since the other calls, particularly `summary()` is mocked out it is not being tested correctly, instead we're testing a mock

If you can, remove the mock for report and mock out the file handling things instead

File: convert2rhel/subscription.py
Comment (hosekadam) [reply convert2rhel/subscription.py:204]: Yeah, I'll add the unit tests now :)

File: convert2rhel/unit_tests/main_test.py
Comment (hosekadam) [reply convert2rhel/unit_tests/main_test.py:141]: @SpyTec I was thinking about it like the `summary()` is being tested elsewhere and here I just need to test if the appropriate log is printed in the right moment and also report is generated when needed (and called with right values). 

But I think I can change that, it won't be a big problem I believe.

**Comment by hosekadam** [comment]
- added the unit test for handling exception in `register_system()`
- added condition to restore the os-release file only in the first attempt. In the next attempts the file is already restored
- removed check if the backup is present and let the possible exception to be raised. Without it, it won’t be handled by the BackupController/something else, and we won’t know what happened. It will be just captured in the log somewhere during the conversion
- added unit tests for restoring only in the first attempt of registering the system
- removed one case of unit test when the file for restoring wasn't present. Now this raises exception
- fixed the comments based on review

File: convert2rhel/unit_tests/main_test.py
Comment (SpyTec) [reply convert2rhel/unit_tests/main_test.py]: Testing logs is fine but we're supposed to test that the function works, not just that it calls what we want. Better to mock out files IMO

File: convert2rhel/backup/__init__.py
Comment (SpyTec) [inline convert2rhel/backup/__init__.py:1]: We probably want to add  `rollback=True` to the restore() method as it's overridden by other classes

**Review by SpyTec — APPROVED** [review]
LGTM, one small thin

> Fixed the problem mentioned in [issues.redhat.com/browse/RHELC-1516](https://issues.redhat.com/browse/RHELC-1516)

Do you mean this PR also fixes that issue? If so please update the PR description and title

File: convert2rhel/backup/__init__.py
Comment (hosekadam) [reply convert2rhel/backup/__init__.py:1]: About the RHELC-1516 I'm not sure which of the possibilities I mentioned in the Jira is better. I added this topic to today's tech cabal.

> We probably want to add rollback=True to the restore() method as it's overridden by other classes
I'm not sure what you mean. the `rollback=True` is already present in the `RestorableFile` (I don't see any other restorable which would need that) and works like a charm.

Just about the integration tests, should I add them? Or will you @danmyway take care of it?

Comment (SpyTec) [reply convert2rhel/backup/__init__.py:1]: This one
image

Comment (hosekadam) [reply convert2rhel/backup/__init__.py:1]: Discussed with @SpyTec and we won't be changing it in this PR

**Comment by SpyTec** [comment]
Has merge conflicts

**Comment by hosekadam** [comment]
Solved and rebased :)

File: convert2rhel/main.py
Comment (r0x0d) [reply convert2rhel/main.py:420]: The pre_conversion_report needs to deindent one level... for some reason making it throught the UI didn't worked

Comment (hosekadam) [reply convert2rhel/main.py:420]: Agree, it looks better

**Comment by hosekadam** [comment]
- rebased, solved conflicts
- fixed the `"` symbol in the one of the comments
- removed `elif` and deindented `report.pre_conversion_report` in `provide_status_after_rollback`

File: convert2rhel/backup/certs.py
Comment (r0x0d) [inline convert2rhel/backup/certs.py:151]:
```suggestion
        .. warning::
            Exceptions are not handled and left for handling by the calling code.
```

Comment (r0x0d) [reply convert2rhel/backup/certs.py:151]: Deindent the waring section

File: convert2rhel/backup/files.py
Comment (r0x0d) [inline convert2rhel/backup/files.py:183]:
```suggestion
        .. warning::
            Exceptions are not handled and left for handling by the calling code.
```

File: convert2rhel/subscription.py
Comment (r0x0d) [inline convert2rhel/subscription.py:258]: Maybe we should use a different id and tittle? We are not subscribing the system at this point, only trying to restore the os_release file. 

The registration comes in the line 261

Comment (r0x0d) [reply convert2rhel/subscription.py:258]: I mean, the procedure is for registering the system, but at this specific step we are not actually registering

**Review by r0x0d — COMMENTED** [review]
Looks good! Only one question though.

File: convert2rhel/subscription.py
Comment (hosekadam) [reply convert2rhel/subscription.py:258]: I was thinking about that, and it resulted in this. I see it that we are trying to subscribe the system and this is part of the process. The problem is then specified in the description and diagnostic. 

But if we would find a better name, I will change that, no problem. I'm not sure which, I felt like `FAILED_TO_RESTORE_OSRELEASE` is a bit misleading.... But I'm not sure there, the naming is the hardest part sometimes :smile:

Comment (SpyTec) [reply convert2rhel/subscription.py:258]: This is a big question and I feel like we need a wiki page or something with all of these things, when to use what, and how each is triggered etc. At the moment it's just "this feels good" but it can lead to confusing things for the customer

Since this prevents the system from being registering it still fails to register, so I feel like this error is still accurate

**Comment by hosekadam** [comment]
Fixed the indentation and rebased

**Comment by hosekadam** [comment]
@danmyway Do we need new integration tests for this or do we have it covered?

**Comment by danmyway** [comment]
Not that critical from my POV, we can add it in a separate batch @hosekadam 
I will just run the tier0, if it passes (apart from the offline conversion test on Alma 8.9) we can merge this

**Comment by Venefilyn** [comment]
LGTM. @hosekadam please update the Jira issues to Dev Complete when you can!

**Pull request merged**
